source/plugins/go/src/ingestion_token_utils.go

```diff
@@ -0,0 +1,437 @@
+	Ingestionauthtoken string `json:"ingestionAuthToken"`
+}
+
+func getAccessTokenFromIMDS() (string, int64, error) {
+	Log("Info getAccessTokenFromIMDS: start")
+	useIMDSTokenProxyEndPoint := os.Getenv("USE_IMDS_TOKEN_PROXY_END_POINT")
+	imdsAccessToken := ""
+	var responseBytes []byte
+	var err error
+
+	if (useIMDSTokenProxyEndPoint != "" && strings.Compare(strings.ToLower(useIMDSTokenProxyEndPoint), "true") == 0) {
+		Log("Info Reading IMDS Access Token from IMDS Token proxy endpoint")
+		MCS_ENDPOINT := os.Getenv("MCS_ENDPOINT")
+		var msi_endpoint *url.URL
+		msi_endpoint, err := url.Parse("http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&resource=https://" + MCS_ENDPOINT + "/")
+		if err != nil {
+			Log("getAccessTokenFromIMDS: Error creating IMDS endpoint URL: %s", err.Error())
+			return imdsAccessToken, 0, err
+		}
+		req, err := http.NewRequest("GET", msi_endpoint.String(), nil)
+		if err != nil {
+			Log("getAccessTokenFromIMDS: Error creating HTTP request: %s", err.Error())
+			return imdsAccessToken, 0, err
+		}
+		req.Header.Add("Metadata", "true")
+
+		//IMDS endpoint nonroutable endpoint and requests doesnt go through proxy hence using separate client
+		httpClient := &http.Client{}
+
+		// Call managed services for Azure resources token endpoint
+		var resp *http.Response = nil
+		for retryCount := 0; retryCount < 3; retryCount++ {
+			resp, err = httpClient.Do(req)
+			if err != nil {
+				message := fmt.Sprintf("getAccessTokenFromIMDS: Error calling token endpoint: %s, retryCount: %d", err.Error(), retryCount)
+				Log(message)
+				SendException(message) // send the exception here because this error is not returned. The calling function will send any returned errors to telemetry.
+				continue
+			}
+
+			if resp != nil && resp.Body != nil {
+			  defer resp.Body.Close()
+			}
+
+			Log("getAccessTokenFromIMDS: IMDS Response Status: %d, retryCount: %d", resp.StatusCode, retryCount)
+			if resp.StatusCode != 200 {
+				message := fmt.Sprintf("getAccessTokenFromIMDS: IMDS Request failed with an error code : %d", resp.StatusCode)
+				Log(message)
+				SendException(message)
+				continue
+			}
+			break // call succeeded, don't retry any more
+		}
+		if resp == nil {
+			Log("getAccessTokenFromIMDS: IMDS Request ran out of retries")
+			return imdsAccessToken, 0, err
+		}
+
+		// Pull out response body
+		responseBytes, err = ioutil.ReadAll(resp.Body)
+		if err != nil {
+			Log("getAccessTokenFromIMDS: Error reading response body : %s", err.Error())
+			return imdsAccessToken, 0, err
+		}
+
+	} else {
+		Log("Info Reading IMDS Access Token from file : %s", IMDSTokenPathForWindows)
+		if _, err = os.Stat(IMDSTokenPathForWindows); os.IsNotExist(err) {
+			Log("getAccessTokenFromIMDS: IMDS token file doesnt exist: %s", err.Error())
+			return imdsAccessToken, 0, err
+		}
+		//adding retries incase if we ended up reading the token file while the token file being written
```

Comment:
A better approach would be to use mutex instead of retrying when the file is being written.

Reply:
How will a mutex help here? (My understanding is the conflict comes from kuberentes updating the secret-file-mount while ingestion_token_utils.go is reading it)